Add support for throwing exceptions

darfink opened the pull request:
I don't really expect this to be merged as is, but I lacked this functionality in my own pet project.
I've implemented support for [`v8::Isolate::ThrowException`](http://v8.paulfryzel.com/docs/master/classv8_1_1_isolate.html#ad2d4dd42267c374a37b93df403ef11ac).

dflemstr commented:
Hey, thanks for the PR, I've been wanting this myself several times.

I was thinking that it would be nice to be able to use Rust idiomatics such as `try!`, so maybe this could be implemented by letting closures return `Result<Value, Value>` or something like that? Is that too confusing?

I am not super worried about breaking backwards compatibility at this stage, so I wouldn't worry about that.

darfink commented:
Interesting. That is more or less exactly how I've wrapped my closures in my own project, and I've felt like it resulted in very elegant and *Rusty™* code. I'll update the implementation using the aforementioned implementation, to allow easy evaluation of the API.

dflemstr commented:
Yeah I think so, sometimes it's useful to throw richer objects. Also, I don't think V8 exposes an `Exception` type anyway really (only when catching things)

darfink commented:
I modified the callback to use `Result<Value, Value>` as return value. I also introduced a type for it (`FunctionCallback`) because I felt the type was becoming very arduous.

dflemstr commented on src/isolate.rs:
```diff
@@ -102,6 +102,19 @@ impl Isolate {
+    /// When an exception has been scheduled it is illegal to invoke any
+    /// JavaScript operation; the caller must return immediately and only after
+    /// the exception has been handled does it become legal to invoke JavaScript
+    /// operations.
```
Does this violate Rust safety? Would it be possible to make this function private or something?

darfink replied:
I actually thought about that as well, and I agree, it should be private. I don't know the cleanest way to do this. I could just move the function implementation to `util.rs`, but you may dislike that the `Isolate` API calls become fragmented?

dflemstr commented on src/isolate.rs:
```diff
@@ -102,6 +102,19 @@ impl Isolate {
+    /// JavaScript operation; the caller must return immediately and only after
+    /// the exception has been handled does it become legal to invoke JavaScript
+    /// operations.
+    pub fn throw_exception(&self, exception: &::value::Value) -> ::value::Value {
```
Interesting to import `Value` using an absolute path here, I don't mind so much but it doesn't match the rest of the code...

dflemstr commented:
Okay this looks good :+1:

dflemstr commented:
Released as 0.8.0
